File: typescript-basic-to-advance/type-systems/type-system.java
import java.util.*;
import java.lang.*;


class Person {
  public String name;
}

class Employee extends Person {
  public int price;
}

class Manager extends Employee { }

class Product {
  public String name;
  public String price;
}

class Main {
  public static void main(String args[]) {
    fun(new Product()); // no error
    fun(new Employee()); // error: incompatible types: Employee cannot be converted to Product

  }
  public static void fun(Product a) {}
}

File: typescript-basic-to-advance/type-systems/type-system.ts
// Nominal Type Checking
// Examples: C++, Java, and Swift

// Refer type-system.java for eample


// Static type checking
class Person {
  public name: string;
}

class Employee extends Person {
  public price: number;
}

class Product {
  public name: string;
  public price: number;
}

function foo (a: Product): void {}

// No error as structure is same
foo(new Employee())